File: shamelagpt-android/app/src/main/java/com/shamelagpt/android/data/mapper/MessageMapper.kt
package com.shamelagpt.android.data.mapper

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.shamelagpt.android.data.local.entity.MessageEntity
import com.shamelagpt.android.domain.model.Message
import com.shamelagpt.android.domain.model.Source

/**
 * Converts MessageEntity to domain model Message.
 *
 * @return Message domain model
 */
fun MessageEntity.toDomain(): Message {
    val sourceList = sources?.let { jsonString ->
        try {
            val gson = Gson()
            val listType = object : TypeToken<List<Source>>() {}.type
            gson.fromJson<List<Source>>(jsonString, listType)
        } catch (e: Exception) {
            null
        }
    }

    return Message(
        id = id,
        content = content,
        isUserMessage = isUserMessage,
        timestamp = timestamp,
        sources = sourceList,
        imageData = imageData,
        detectedLanguage = detectedLanguage,
        isFactCheckMessage = isFactCheckMessage
    )
}

/**
 * Converts domain model Message to MessageEntity.
 *
 * @param conversationId ID of the conversation this message belongs to
 * @return MessageEntity for database storage
 */
fun Message.toEntity(conversationId: String): MessageEntity {
    val sourcesJson = sources?.let { sourceList ->
        try {
            val gson = Gson()
            gson.toJson(sourceList)
        } catch (e: Exception) {
            null
        }
    }

    return MessageEntity(
        id = id,
        conversationId = conversationId,
        content = content,
        isUserMessage = isUserMessage,
        timestamp = timestamp,
        sources = sourcesJson,
        imageData = imageData,
        detectedLanguage = detectedLanguage,
        isFactCheckMessage = isFactCheckMessage
    )
}

File: shamelagpt-ios/shamelagpt/Data/Repositories/MessageMapper.swift
//
//  MessageMapper.swift
//  ShamelaGPT
//
//  Created by Ameed Khalid on 05/11/2025.
//

import Foundation

/// Maps between MessageEntity (Core Data) and Message (Domain Model)
struct MessageMapper {

    // MARK: - Entity to Domain Model

    /// Converts a MessageEntity to a Message domain model
    /// - Parameter entity: The MessageEntity to convert
    /// - Returns: A Message domain model
    static func toDomainModel(_ entity: MessageEntity) -> Message {
        let sources = parseSources(from: entity.sources)

        return Message(
            id: entity.id ?? UUID().uuidString,
            conversationId: entity.conversationId ?? "",
            content: entity.content ?? "",
            isUserMessage: entity.isUserMessage,
            timestamp: entity.timestamp ?? Date(),
            sources: sources,
            imageData: entity.imageData,
            detectedLanguage: entity.detectedLanguage,
            isFactCheckMessage: entity.isFactCheckMessage
        )
    }

    /// Converts an array of MessageEntity to an array of Message domain models
    /// - Parameter entities: The array of MessageEntity to convert
    /// - Returns: An array of Message domain models
    static func toDomainModels(_ entities: [MessageEntity]) -> [Message] {
        return entities.map { toDomainModel($0) }
    }

    // MARK: - Domain Model to JSON

    /// Converts sources array to JSON string for storage
    /// - Parameter sources: Array of Source objects
    /// - Returns: JSON string representation, or nil if empty or encoding fails
    static func sourcesToJSON(_ sources: [Source]) -> String? {
        guard !sources.isEmpty else { return nil }

        let encoder = JSONEncoder()
        encoder.outputFormatting = .sortedKeys

        do {
            let data = try encoder.encode(sources)
            return String(data: data, encoding: .utf8)
        } catch {
            print("Error encoding sources to JSON: \(error)")
            return nil
        }
    }

    // MARK: - JSON to Domain Model

    /// Parses sources from JSON string
    /// - Parameter jsonString: JSON string containing sources
    /// - Returns: Array of Source objects, empty if parsing fails
    private static func parseSources(from jsonString: String?) -> [Source] {
        guard let jsonString = jsonString,
              let data = jsonString.data(using: .utf8) else {
            return []
        }

        let decoder = JSONDecoder()

        do {
            return try decoder.decode([Source].self, from: data)
        } catch {
            print("Error decoding sources from JSON: \(error)")
            return []
        }
    }
}
